### PR TITLE
ref(datasets): Move time group magic column to `TimeSeriesDataset`

### PR DESCRIPTION
**Pull request by tkaemming**

### CODE REVIEW

**Review by tkaemming — COMMENTED** [review]
This is related to SNS-207.

I'd like to eventually also make query splitting part of this abstraction (at least [`time_split`](https://github.com/getsentry/snuba/blob/4b36a882c3ed167dbe84dba9e17eb4b2ec9e5352/snuba/split.py#L38)), and as well as making [the time series specific values in the schema](https://github.com/getsentry/snuba/blob/4b36a882c3ed167dbe84dba9e17eb4b2ec9e5352/snuba/schemas.py#L68-L81) only be added to queries to the dataset(s) where they are pertinent, but that is going to take some (probably a lot) more work. The fact that the [SDK stats query schema has different defaults](https://github.com/getsentry/snuba/blob/4b36a882c3ed167dbe84dba9e17eb4b2ec9e5352/snuba/schemas.py#L11-L24) than the generic query makes this more difficult. (In general, the special casing of the SDK stats query has created me a lot of frustration as I've tried to make broader changes here.)

File: snuba/datasets/__init__.py [hunk 1]
```diff
@@ -59,13 +64,35 @@ def column_expr(self, column_name, body):
...
         """
         raise NotImplementedError
+
+
+class TimeSeriesDataset(Dataset):
+    def __init__(self, *args, time_group_columns: Mapping[str, str], **kwargs):
+        super().__init__(*args, **kwargs)
+        # Convenience columns that evaluate to a bucketed time. The bucketing
+        # depends on the granularity parameter.
+        self.__time_group_columns = time_group_columns
```
Comment (tkaemming) [inline snuba/datasets/__init__.py]: Might also make sense to assert that these columns actually exist in the schema.

File: snuba/datasets/__init__.py [hunk 2]
```diff
@@ -59,13 +64,35 @@ def column_expr(self, column_name, body):
...
         """
         Returns the instance of the bulk loader to populate the dataset from an
         external source when present.
         """
         raise NotImplementedError
+
+
+class TimeSeriesDataset(Dataset):
+    def __init__(self, *args, time_group_columns: Mapping[str, str], **kwargs):
+        super().__init__(*args, **kwargs)
```
[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba/datasets/__init__.py]: mindblown

File: snuba/datasets/__init__.py [hunk 3]
```diff
@@ -59,13 +64,35 @@ def column_expr(self, column_name, body):
...
+        # depends on the granularity parameter.
+        self.__time_group_columns = time_group_columns
+
+    def __time_expr(self, column_name: str, granularity: int) -> str:
+        real_column = self.__time_group_columns[column_name]
+        template = {
+            3600: 'toStartOfHour({column})',
+            60: 'toStartOfMinute({column})',
+            86400: 'toDate({column})',
+        }.get(granularity, 'toDateTime(intDiv(toUInt32({column}), {granularity}) * {granularity})')
```
Comment (fpacifici) [inline snuba/datasets/__init__.py]: Would it make sense to limit the minimum allowed granularities per dataset ? Like not allowing someone to group by transactions by 2 5 seconds timeframes ?

File: snuba/datasets/events.py [hunk 4]
```diff
@@ -27,7 +26,7 @@
 NESTED_COL_EXPR_RE = re.compile(r'^(tags|contexts)\[([a-zA-Z0-9_\.:-]+)\]$')
 
 
-class EventsDataset(Dataset):
+class EventsDataset(TimeSeriesDataset):
```
Comment (fpacifici) [inline snuba/datasets/events.py]: I am concerned this can easily become a mixin, it seems to me we are going towards a lot of datasets customization that each dataset will pick features from different places. (think about a CDC dataset with timeseries support, with delegating event processor, and tag promotion).
Would it make sense to make TimeSeriesDataset either a mixin (then we would need to define the semantics of column_expr when two mixins define it)
Or, better in my opinion, go for composition and make TimeSeriesSupport a class that extends "ColumnExpression" that exposes the column_expr method. The column_expr method on the dataset would delegate the column processing to the ColumnExpression classes defined in the dataset config ?

File: snuba/datasets/__init__.py (hunk 3 above)
Comment (tkaemming) [reply snuba/datasets/__init__.py]: No strong opinion really, I don't think it really makes a performance difference but I also don't think there's a reason anybody should need/want extreme flexibility either here.

File: snuba/datasets/events.py (hunk 4 above)
Comment (tkaemming) [reply snuba/datasets/events.py]: > I am concerned this can easily become a mixin, it seems to me we are going towards a lot of datasets customization that each dataset will pick features from different places. (think about a CDC dataset with timeseries support, with delegating event processor, and tag promotion).

I share this opinion.

> Would it make sense to make TimeSeriesDataset either a mixin (then we would need to define the semantics of column_expr when two mixins define it)

I'd like to avoid mixins (and diamond inheritance) for these reasons as well.

> Or, better in my opinion, go for composition and make TimeSeriesSupport a class that extends "ColumnExpression" that exposes the column_expr method. The column_expr method on the dataset would delegate the column processing to the ColumnExpression classes defined in the dataset config ?

I think that I agree this should be the longer-term approach that we take here, but I'm a little hesitant to make that broad of a change until it's necessary and/or we have a better idea about what other kinds of query-time functionality needs to be encapsulated in the time-series dataset. (I am a little concerned that using several different column expression extensions on the same dataset will lead to the same issues as mixins or multiple inheritance, but at least in that case the scope is narrower.)

In other words — I'm pretty confident we'll end up doing this (or something pretty similar), I just don't know as we *need* to right now. Maybe a compromise would be to make a ticket to follow up on this to make sure that we address it (even if not absolutely necessary) if/when we have some time to circle back and make general improvements; otherwise we'll just tackle it if/when it's a necessity?

Comment (fpacifici) [reply snuba/datasets/events.py]: sure, mind filing the ticket ?

**Review by fpacifici — APPROVED** [review]

File: snuba/datasets/events.py (hunk 4 above)
Comment (tkaemming) [reply snuba/datasets/events.py]: This is recorded at SNS-250 now.
